File: crates/clipmilld/src/inputs.rs
//! Finding the documents a derivation stage was told to read.
//!
//! The four model-free stages that read published artifacts — the evidence
//! index, discovery, ranking, and the analysis fan-in — reach their inputs by
//! two routes that arrive at one place. Submitted as a standalone job, what they
//! read was published by *earlier jobs*, so the plan declares its addresses.
//! Run inside the analyze DAG, those same documents are the outputs of tasks in
//! the same plan, so a dependency carries them. Either way the daemon delivers
//! one list on the lease, and this module reads it.
//!
//! That the two routes converge before this point is the property that matters.
//! A stage that resolved a payload address on one route and a dependency on the
//! other would compute two keys for one piece of work — and the artifact key
//! covers the input list, so those two keys would be two addresses for one
//! observation. Nothing in a content-addressed store can notice that afterwards.
//!
//! Inputs are matched to what a stage asked for by the artifact kind each one's
//! own manifest declares, never by position. A plan that declared a transcript
//! where an index belongs, or a dependency list reordered upstream, is refused
//! rather than parsed into something plausible.
//!
//! Everything here needs a live artifact store, so it is exercised end to end
//! by `gate-evidence`, `gate-discovery`, and `gate-ranking` against a real
//! daemon rather than by unit tests against a mock that would only prove the
//! mock agrees with itself.

use std::collections::BTreeMap;

use clipmill_artifacts::ArtifactLease;
use clipmill_core::ArtifactId;

use crate::{
    artifacts::ArtifactHandle,
    jobs::{LeasedTask, TaskExecutionError},
    media,
};

/// One document a stage needs, named by the artifact kind it must be.
#[derive(Clone, Copy, Debug)]
pub(crate) struct Wanted {
    pub kind: &'static str,
    /// A missing optional input is an absence the stage reports rather than a
    /// failure — a source with no video has no shot cuts, and that is a
    /// different document from one whose shot detection was never run.
    pub required: bool,
}

impl Wanted {
    pub(crate) fn required(kind: &'static str) -> Self {
        Self {
            kind,
            required: true,
        }
    }

    pub(crate) fn optional(kind: &'static str) -> Self {
        Self {
            kind,
            required: false,
        }
    }
}

/// The documents a stage may read, opened and verified.
pub(crate) struct Resolved {
    /// In the order the caller asked for them, so the artifact key does not
    /// depend on a map's iteration order.
    order: Vec<&'static str>,
    found: BTreeMap<&'static str, (ArtifactId, ArtifactLease)>,
}

/// Open everything the stage was told to read.
pub(crate) async fn resolve(
    artifacts: &ArtifactHandle,
    task: &LeasedTask,
    wanted: &[Wanted],
) -> Result<Resolved, TaskExecutionError> {
    // Everything the lease delivered, indexed by the kind each one declares.
    // Opened once: a stage with three inputs would otherwise open the same
    // artifact three times to ask what it is.
    let mut by_kind: BTreeMap<String, (ArtifactId, ArtifactLease)> = BTreeMap::new();
    for artifact_id in &task.input_artifact_ids {
        let lease = artifacts
            .open(*artifact_id)
            .await
            .map_err(|error| TaskExecutionError::transient(error.to_string()))?;
        let kind = lease.kind().to_owned();
        if by_kind
            .insert(kind.clone(), (*artifact_id, lease))
            .is_some()
        {
            // Two inputs of one kind leave the stage no basis for choosing,
            // and picking either would be picking by position.
            return Err(TaskExecutionError::deterministic(format!(
                "this task was given two {kind} inputs and names no choice"
            )));
        }
    }

    let mut found = BTreeMap::new();
    let mut order = Vec::new();
    for entry in wanted {
        order.push(entry.kind);
        match by_kind.remove(entry.kind) {
            Some(pair) => {
                found.insert(entry.kind, pair);
            }
            None if entry.required => {
                return Err(TaskExecutionError::deterministic(format!(
                    "this task has no {} to read",
                    entry.kind
                )));
            }
            None => {}
        }
    }
    // An input nobody asked for is a plan that believes this stage reads
    // something it does not. Refused rather than ignored: the artifact key
    // covers only what was read, so a stage silently handed a fourth document
    // would key as though it had three.
    if let Some((kind, _)) = by_kind.into_iter().next() {
        return Err(TaskExecutionError::deterministic(format!(
            "this task was given a {kind} it does not read"
        )));
    }
    Ok(Resolved { order, found })
}

impl Resolved {
    /// Read one document out of the artifact that carries it.
    pub(crate) fn read<T: serde::de::DeserializeOwned>(
        &self,
        kind: &'static str,
        file: &str,
    ) -> Result<T, TaskExecutionError> {
        let (_, lease) = self.found.get(kind).ok_or_else(|| {
            TaskExecutionError::deterministic(format!("this task has no {kind} to read"))
        })?;
        media::read_artifact_document(lease, file)
    }

    /// The same, for an input that may legitimately be absent.
    pub(crate) fn read_optional<T: serde::de::DeserializeOwned>(
        &self,
        kind: &'static str,
        file: &str,
    ) -> Result<Option<T>, TaskExecutionError> {
        match self.found.get(kind) {
            None => Ok(None),
            Some((_, lease)) => media::read_artifact_document(lease, file).map(Some),
        }
    }

    /// The address of one input, for the document that records what it read.
    pub(crate) fn address(&self, kind: &'static str) -> Result<String, TaskExecutionError> {
        self.found
            .get(kind)
            .map(|(artifact_id, _)| artifact_id.to_string())
            .ok_or_else(|| {
                TaskExecutionError::deterministic(format!("this task has no {kind} to read"))
            })
    }

    /// The address of an input that may be absent.
    pub(crate) fn optional_address(&self, kind: &'static str) -> Option<String> {
        self.found
            .get(kind)
            .map(|(artifact_id, _)| artifact_id.to_string())
    }

    /// Every address this stage read, in the order they were asked for.
    ///
    /// This is what the artifact key covers, so a stage that read three
    /// documents cannot be served a cache entry computed from two.
    pub(crate) fn addresses(&self) -> Vec<ArtifactId> {
        self.order
            .iter()
            .filter_map(|kind| self.found.get(kind).map(|(artifact_id, _)| *artifact_id))
            .collect()
    }
}

File: workers/sdk/src/clipmill_worker_sdk/inputs.py
"""Finding the artifacts a lease delivered, by what each one is.

A stage's inputs reach it by one of two routes. Submitted as a standalone job,
what it reads was published by earlier jobs, so the plan declares their content
addresses. Run inside a larger plan, those same artifacts are the outputs of
tasks it depends on, so a dependency carries them. Either way the daemon
delivers one list on the lease, and this is how a worker reads it.

That the routes converge before the worker sees them is the point. A stage that
resolved an address out of its own payload on one route and a dependency on the
other would compute two artifact keys for one piece of work — and a
content-addressed store cannot notice two addresses for one observation after
the fact. It is also the only route that *can* work: a worker may open exactly
the artifacts its lease named, so an address that travelled only in the payload
named something the worker was forbidden to read.

Inputs are matched by the artifact kind each one's own manifest declares, never
by position. A stage handed two artifacts of one kind has no basis for choosing
between them and says so, rather than taking the first and being right most of
the time.
"""

from __future__ import annotations

from dataclasses import dataclass

from .artifacts import ArtifactVerificationError, VerifiedArtifact


class MissingInputError(RuntimeError):
    """A lease did not deliver an artifact the stage cannot run without."""


@dataclass(frozen=True, slots=True)
class ResolvedInput:
    """One artifact this lease delivered, opened and verified."""

    kind: str
    artifact_id: str
    artifact: VerifiedArtifact


class LeaseInputs:
    """Everything a lease delivered, indexed by kind.

    Opened once. A stage with three inputs would otherwise open the same
    artifacts repeatedly just to ask what they are, and every open re-hashes the
    payload it verifies.
    """

    def __init__(self, context) -> None:
        self._by_kind: dict[str, ResolvedInput] = {}
        for artifact_id in context.lease.input_artifact_ids:
            artifact = context.open_artifact(artifact_id)
            if artifact.kind in self._by_kind:
                raise MissingInputError(
                    f"this lease delivered two {artifact.kind} inputs and names no choice"
                )
            self._by_kind[artifact.kind] = ResolvedInput(
                kind=artifact.kind, artifact_id=artifact_id, artifact=artifact
            )

    def require(self, kind: str) -> ResolvedInput:
        """The one input of this kind, or a refusal naming what was delivered.

        A refusal rather than a fallback: a stage that guessed which of several
        inputs was meant is how a worker ends up reading last week's artifact and
        publishing it under this week's key.
        """

        found = self._by_kind.get(kind)
        if found is None:
            delivered = ", ".join(sorted(self._by_kind)) or "nothing"
            raise MissingInputError(f"this lease delivered {delivered}, not a {kind}")
        return found

    def optional(self, kind: str) -> ResolvedInput | None:
        """The same, for an input that may legitimately be absent.

        Absent is a fact the stage reports, not a failure it hides: a source with
        no video has no shot cuts, and that is a different observation from one
        whose shot detection was never run.
        """

        return self._by_kind.get(kind)

    def kinds(self) -> tuple[str, ...]:
        return tuple(sorted(self._by_kind))


def require_input(context, kind: str) -> ResolvedInput:
    """One input of a kind, for a stage that reads exactly one artifact.

    The `ArtifactVerificationError` a bad store raises is not caught here. It
    means the object the daemon pointed at is not what its manifest says, which
    is neither this stage's mistake nor something it can work around.
    """

    return LeaseInputs(context).require(kind)


__all__ = [
    "ArtifactVerificationError",
    "LeaseInputs",
    "MissingInputError",
    "ResolvedInput",
    "require_input",
]
